src/rust/protocol/src/memcache/wire/request/parse.rs

```diff
@@ -0,0 +1,368 @@
+                keys.push(buffer[previous..key_end].to_vec().into_boxed_slice());
+                previous = key_end + 1;
+            } else {
+                keys.push(buffer[previous..line_end].to_vec().into_boxed_slice());
+                break;
+            }
+        } else {
+            keys.push(buffer[previous..line_end].to_vec().into_boxed_slice());
+            break;
+        }
+    }
+
+    let consumed = line_end + CRLF.len();
+
+    let message = MemcacheRequest::Get {
+        keys: keys.into_boxed_slice(),
+    };
+    Ok(ParseOk { message, consumed })
+}
+
+fn parse_gets(buffer: &[u8]) -> Result<ParseOk<MemcacheRequest>, ParseError> {
+    let request = parse_get(buffer)?;
+    let consumed = request.consumed();
+    let message = if let MemcacheRequest::Get { keys } = request.into_inner() {
+        MemcacheRequest::Gets { keys }
+    } else {
+        unreachable!()
+    };
+
+    Ok(ParseOk { message, consumed })
+}
+
+fn parse_set(buffer: &[u8]) -> Result<ParseOk<MemcacheRequest>, ParseError> {
+    let mut single_byte = buffer.windows(1);
+    if let Some(cmd_end) = single_byte.position(|w| w == b" ") {
+        // key
+        let key_end = single_byte
+            .position(|w| w == b" ")
+            .ok_or(ParseError::Invalid)?
+            + cmd_end
+            + 1;
+
+        // flags
+        let flags_end = single_byte
+            .position(|w| w == b" ")
+            .ok_or(ParseError::Invalid)?
+            + key_end
+            + 1;
+        let flags_str = std::str::from_utf8(&buffer[(key_end + 1)..flags_end])
+            .map_err(|_| ParseError::Invalid)?;
+        let flags = flags_str.parse().map_err(|_| ParseError::Invalid)?;
+
+        // expiry
+        let expiry_end = single_byte
+            .position(|w| w == b" ")
+            .ok_or(ParseError::Invalid)?
+            + flags_end
+            + 1;
+        let expiry_str = std::str::from_utf8(&buffer[(flags_end + 1)..expiry_end])
+            .map_err(|_| ParseError::Invalid)?;
+        let expiry = expiry_str.parse().map_err(|_| ParseError::Invalid)?;
+
+        // now it gets tricky, we either have "[bytes] noreply\r\n" or "[bytes]\r\n"
+        let mut double_byte = buffer.windows(CRLF.len());
+        let mut noreply = false;
+
+        // get the position of the next space and first CRLF
+        // we already know we must have a CRLF if we're parsing a set request
+        let next_space = single_byte
+            .position(|w| w == b" ")
+            .map(|v| v + expiry_end + 1);
+        let first_crlf = double_byte
+            .position(|w| w == CRLF.as_bytes())
+            .ok_or(ParseError::Invalid)?;
+
+        let bytes_end = if let Some(next_space) = next_space {
+            // if we have both, bytes_end is before the earlier of the two
+            if next_space < first_crlf {
+                // validate that noreply isn't malformed
+                if &buffer[(next_space + 1)..(first_crlf)] == NOREPLY.as_bytes() {
```

Comment:
Years ago (~2015) I got tripped during canary by commands that were slightly malformatted due to a similarly strict parser, the command was something like `set apple 0 0 6 \r\nOrange\r\n`. Notice there is an extra whitespace after value length, `6`. I think this type of error is not super rare because programmers often print a whitespace after each field and not special-casing the last field before the line ends. So to be defensive against this rather benign implementation error, you may want to allow at least one whitespace after the last field.

Reply:
Will add a test and allow that. Interesting!

Reply:
This is now fixed and even covered by tests! =)